src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java

```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
-        {
-            // force prefetch before starting record time
-            boolean builkFetch = returnedObject.getBulkFetch();
-            returnedObject.setBulkFetch(false);
-            returnedObject.getNodeRef(returnedObject.length() - 1);
-            returnedObject.setBulkFetch(builkFetch);
+            return resultSet;
         }
 
-        // record the start time
-        long startTimeMillis = System.currentTimeMillis();
-        // set the default, unlimited resultset type
-        filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(returnedObject.getResultSetMetaData().getLimitedBy(), PermissionEvaluationMode.EAGER, returnedObject.getResultSetMetaData()
-                .getSearchParameters()));
+        return filterMaxCount(maxSize, resultSet);
+    }
 
-        for (int i = 0; i < returnedObject.length(); i++)
+    private ResultSet decideOld(Authentication authentication, Object object, ConfigAttributeDefinition config, ResultSet returnedObject) throws AccessDeniedException
+    {
+        if (returnedObject == null)
         {
-            long currentTimeMillis = System.currentTimeMillis();
-            if (i >= maxChecks)
-            {
-                log.warn("maxChecks exceeded (" + maxChecks + ")", new Exception("Back Trace"));
-                filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.NUMBER_OF_PERMISSION_EVALUATIONS, PermissionEvaluationMode.EAGER, returnedObject
-                        .getResultSetMetaData().getSearchParameters()));
-                break;
-            }
-            else if ((currentTimeMillis - startTimeMillis) > maxCheckTime)
-            {
-                log.warn("maxCheckTime exceeded (" + (currentTimeMillis - startTimeMillis) + " milliseconds)", new Exception("Back Trace"));
-                filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.NUMBER_OF_PERMISSION_EVALUATIONS, PermissionEvaluationMode.EAGER, returnedObject
-                        .getResultSetMetaData().getSearchParameters()));
-                break;
-            }
-
-            // All permission checks must pass
-            filteringResultSet.setIncluded(i, true);
+            return null;
+        }
 
-            for (ConfigAttributeDefintion cad : supportedDefinitions)
-            {
-                NodeRef testNodeRef = null;
-                if (cad.typeString.equals(AFTER_ACL_NODE))
-                {
-                    testNodeRef = returnedObject.getNodeRef(i);
-                }
-                else if (cad.typeString.equals(AFTER_ACL_PARENT))
-                {
-                    testNodeRef = returnedObject.getChildAssocRef(i).getParentRef();
-                }
+        List<ConfigAttributeDefintion> supportedDefinitions = extractSupportedDefinitions(config);
+        ResultSet resultSet = supportedDefinitions.size() == 0? returnedObject: decidePermissions(returnedObject, supportedDefinitions);

```

Comment:
Rather than `.size() == 0` we can use `.isEmpty()`, which looks a tiny bit neater.